cmd/zpool/zpool_main.c

```diff
@@ -2844,12 +3441,262 @@ get_timestamp_arg(char c)
 }
 
 /*
- * zpool iostat [-gLPv] [-T d|u] [pool] ... [interval [count]]
+ * Return stat flags that are supported by all pools.  "data" should be
+ * initialized to all 0xFFs before running.  It will get ANDed down until
+ * only the flags that are supported on all pools remain.
+ */
+static int
+get_stat_flags_cb(zpool_handle_t *zhp, void *data)
+{
+	uint64_t *mask = data;
+	nvlist_t *config, *nvroot;
+	uint64_t flags = 0;
+	int i, j;
+
+	/*
+	 * Lookup table for iostat flags to nvlist names.  Basically a list
+	 * of all the nvlists a flag requires.
+	 */
+	static const char *type_to_nvlist[IOS_COUNT][10] = {
+		[IOS_DEFAULT] = {ZPOOL_CONFIG_VDEV_STATS, NULL},
+		[IOS_L_HISTO] = {ZPOOL_CONFIG_TOT_R_LAT_HISTO,
+		    ZPOOL_CONFIG_TOT_W_LAT_HISTO,
+		    ZPOOL_CONFIG_DISK_R_LAT_HISTO,
+		    ZPOOL_CONFIG_DISK_W_LAT_HISTO,
+		    ZPOOL_CONFIG_SYNC_R_LAT_HISTO,
+		    ZPOOL_CONFIG_SYNC_W_LAT_HISTO,
+		    ZPOOL_CONFIG_ASYNC_R_LAT_HISTO,
+		    ZPOOL_CONFIG_ASYNC_W_LAT_HISTO,
+		    ZPOOL_CONFIG_SCRUB_LAT_HISTO, NULL},
+		[IOS_LATENCY] = {ZPOOL_CONFIG_TOT_R_LAT_HISTO,
+		    ZPOOL_CONFIG_TOT_W_LAT_HISTO,
+		    ZPOOL_CONFIG_DISK_R_LAT_HISTO,
+		    ZPOOL_CONFIG_DISK_W_LAT_HISTO, NULL},
+		[IOS_QUEUES] = {ZPOOL_CONFIG_QUEUE_ACTIVE, NULL}
+	};
+
+	config = zpool_get_config(zhp, NULL);
+	verify(nvlist_lookup_nvlist(config, ZPOOL_CONFIG_VDEV_TREE,
+	    &nvroot) == 0);
+
+	/* For each type, make sure all its nvlist entries are supported */
+	for (j = 0; j < ARRAY_SIZE(type_to_nvlist); j++) {
+		if (!type_to_nvlist[j][0])
+			continue;
+
+		/* Start off by assuming the flag is supported, then check */
+		flags |= (1ULL << j);
```

Comment:
I'd like to avoid defining a IOS_ALLFLAGS_M mask if possible.  While it would make the line more readable, it would also be something we would have to update with every new stat added (and I do plan on adding more stats in the not-to-distant future...)